### PR TITLE
Add jitter and extra cache for background processes

### PR DESCRIPTION
**Pull request by aboedo**
Apps opening from push notifications and potentially widgets may cause unnecessary strain on the backend by making calls simultaneously on all devices.

To alleviate the backend strain, we want to introduce a small, random delay to cache updates performed while the app is in the background. This will distribute requests over time so that the backend isn’t hit at once.

We’re also going to be more aggressive with the cache while the app is in the background, moving it from 5 mins to 1 day.

The value of 5 mins will still be used if the app is in the foreground.

Ruleset:

- if app is opened and in the foreground
    - if cache is older than 5 mins
        - update cache
- if app is opened and in the background:
    - if cache is older than one day:
        - update cache (with jitter)

### CODE REVIEW

File: Purchases/Caching/RCDeviceCache.m [hunk 1]
```diff
@@ -8,15 +8,12 @@
...
 #import "RCDeviceCache+Protected.h"
-#import "RCLogUtils.h"
-#import "NSDictionary+RCExtensions.h"
 
 @interface RCDeviceCache ()
 
 @property (nonatomic) NSUserDefaults *userDefaults;
 @property (nonatomic) NSNotificationCenter *notificationCenter;
 @property (nonatomic, nonnull) RCInMemoryCachedObject<RCOfferings *> *offeringsCachedObject;
-@property (nonatomic, nullable) NSDate *purchaserInfoCachesLastUpdated;
```
Comment (aboedo) [inline Purchases/Caching/RCDeviceCache.m:19]: We were storing this value in memory before. This doesn't work well for widgets, since the process for the widget will likely be killed before the next refresh, so it'll always refresh cache.

File: Purchases/Caching/RCDeviceCache.m [hunk 2]
```diff
@@ -8,15 +8,12 @@
 
 #import "RCDeviceCache.h"
 #import "RCDeviceCache+Protected.h"
-#import "RCLogUtils.h"
-#import "NSDictionary+RCExtensions.h"
```
Comment (aboedo) [inline Purchases/Caching/RCDeviceCache.m:12]: unused imports

File: Purchases/Caching/RCDeviceCache.h [hunk 3]
```diff
@@ -33,21 +33,21 @@ NS_ASSUME_NONNULL_BEGIN
 
 - (void)cachePurchaserInfo:(NSData *)data forAppUserID:(NSString *)appUserID;
 
-- (BOOL)isPurchaserInfoCacheStale;
+- (BOOL)isPurchaserInfoCacheStaleForAppUserID:(NSString *)appUserID isAppBackgrounded:(BOOL)isAppBackgrounded;
```
Comment (aboedo) [inline Purchases/Caching/RCDeviceCache.h:36]: not a huge fan of how this API is turning out, it feels overly complicated. But I didn't want to perform a major refactor for this.

File: Purchases/Caching/RCDeviceCache.m [hunk 4]
```diff
@@ -26,11 +23,12 @@ @interface RCDeviceCache ()
 NSString *RCLegacyGeneratedAppUserDefaultsKey = RC_CACHE_KEY_PREFIX @".appUserID";
 NSString *RCAppUserDefaultsKey = RC_CACHE_KEY_PREFIX @".appUserID.new";
 NSString *RCPurchaserInfoAppUserDefaultsKeyBase = RC_CACHE_KEY_PREFIX @".purchaserInfo.";
+NSString *RCPurchaserInfoLastUpdatedKeyBase = RC_CACHE_KEY_PREFIX @".purchaserInfoLastUpdated.";
```
Comment (aboedo) [inline Purchases/Caching/RCDeviceCache.m:26]: new `userDefault` entry

File: Purchases/Caching/RCDeviceCache.m [hunk 5]
```diff
@@ -125,34 +122,55 @@ - (void)cachePurchaserInfo:(NSData *)data forAppUserID:(NSString *)appUserID {
     @synchronized (self) {
         [self.userDefaults setObject:data
                               forKey:[self purchaserInfoUserDefaultCacheKeyForAppUserID:appUserID]];
-        [self setPurchaserInfoCacheTimestampToNow];
+        [self setPurchaserInfoCacheTimestampToNowForAppUserID:appUserID];
```
Comment (aboedo) [inline Purchases/Caching/RCDeviceCache.m:125]: this didn't used to be tied to an `appUserID`, but I tied it now for the sake of consistency.

File: Purchases/Caching/RCDeviceCache.m [hunk 6]
```diff
@@ -163,8 +181,9 @@ - (void)cacheOfferings:(RCOfferings *)offerings {
     [self.offeringsCachedObject cacheInstance:offerings];
 }
 
-- (BOOL)isOfferingsCacheStale {
-    return self.offeringsCachedObject.isCacheStale;
+- (BOOL)isOfferingsCacheStaleWithIsAppBackgrounded:(BOOL)isAppBackgrounded {
```
Comment (aboedo) [inline Purchases/Caching/RCDeviceCache.m:184]: having a different cache might not be strictly necessary for offerings, really - it's unlikely that you'd need offerings while the app is running in the background. can't hurt, though.

File: Purchases/Public/RCPurchases.m [hunk 7]
```diff
@@ -299,7 +299,7 @@ - (instancetype)initWithAppUserID:(nullable NSString *)appUserID
 
         [self.systemInfo isApplicationBackgroundedWithCompletion:^(BOOL isBackgrounded) {
             if (!isBackgrounded) {
-                [self.operationDispatcher dispatchOnWorkerThread:^{
+                [self.operationDispatcher dispatchOnWorkerThreadWithRandomDelay:NO :^{
```
Comment (aboedo) [inline Purchases/Public/RCPurchases.m:302]: the value for `randomDelay` is always `isBackgrounded`, it might make sense to have the `operationDispatcher` depend on systemInfo and get the value without a param

File: Purchases/Public/RCPurchases.m [hunk 8]
```diff
@@ -421,18 +421,20 @@ - (void)resetWithCompletionBlock:(nullable RCReceivePurchaserInfoBlock)completio
...
+        if (infoFromCache) {
+            RCDebugLog(@"Vending purchaserInfo from cache");
+            CALL_IF_SET_ON_MAIN_THREAD(completion, infoFromCache, nil);
+            if ([self.deviceCache isPurchaserInfoCacheStaleForAppUserID:self.appUserID isAppBackgrounded:isAppBackgrounded]) {
+                RCDebugLog(@"Cache is stale, updating caches");
+                [self fetchAndCachePurchaserInfoWithCompletion:nil isAppBackgrounded:isAppBackgrounded];
+            }
+        } else {
+            RCDebugLog(@"No cached purchaser info, fetching");
+            [self fetchAndCachePurchaserInfoWithCompletion:completion isAppBackgrounded:isAppBackgrounded];
```
Comment (aboedo) [inline Purchases/Public/RCPurchases.m:435]: really don't love making these even more complex, but I couldn't think of a way around it given that the isAppBackgrounded method is async.

File: Purchases/Public/RCPurchases.m [hunk 9]
```diff
@@ -821,14 +824,16 @@ - (void)sendCachedPurchaserInfoIfAvailable {
...
+    [self.systemInfo isApplicationBackgroundedWithCompletion:^(BOOL isAppBackgrounded) {
+        if ([self.deviceCache isPurchaserInfoCacheStaleForAppUserID:self.appUserID isAppBackgrounded:isAppBackgrounded]) {
+            RCDebugLog(@"PurchaserInfo cache is stale, updating caches");
+            [self fetchAndCachePurchaserInfoWithCompletion:nil isAppBackgrounded:isAppBackgrounded];
+        }
+        if ([self.deviceCache isOfferingsCacheStaleWithIsAppBackgrounded:isAppBackgrounded]) {
+            RCDebugLog(@"Offerings cache is stale, updating caches");
+            [self updateOfferingsCache:nil isAppBackgrounded:isAppBackgrounded];
+        }
+    }];
```
Comment (aboedo) [inline Purchases/Public/RCPurchases.m:836]: would love to move all this stuff into a new class, but it feels too risky for this particular feature

File: Purchases/Public/RCPurchases.m [hunk 10]
```diff
@@ -863,25 +868,30 @@ - (void)cachePurchaserInfo:(RCPurchaserInfo *)info forAppUserID:(NSString *)appU
...
-                                              [self cachePurchaserInfo:info forAppUserID:appUserID];
-                                              [self sendUpdatedPurchaserInfoToDelegateIfChanged:info];
-                                          } else {
-                                              [self.deviceCache clearPurchaserInfoCacheTimestamp];
-                                          }
-
-                                          CALL_IF_SET_ON_MAIN_THREAD(completion, info, error);
-                                      }];
+    [self.deviceCache setPurchaserInfoCacheTimestampToNowForAppUserID:appUserID];
+    [self.operationDispatcher dispatchOnWorkerThreadWithRandomDelay:isAppBackgrounded :^{
```
Comment (aboedo) [inline Purchases/Public/RCPurchases.m:881]: this API came out a bit weird - note the nameless second param that just starts with `:`

File: Purchases/Public/RCPurchases.m (hunk 7 above)
Comment (vegaro) [reply Purchases/Public/RCPurchases.m:302]: I actually think the param is better. That way we don't add logic to the operationDispatcher that depends on the state of the system. The operation dispatcher only knows what it means to add a random delay, it doesn't care if the app is backgrounded or not. Making `operationDispatcher` depend on the system info also means that you don't have flexibility to do the random delay only for getOfferings or getPurchaserInfo

File: PurchasesTests/Caching/DeviceCacheTests.swift [hunk 11]
```diff
@@ -194,4 +206,41 @@ class DeviceCacheTests: XCTestCase {
...
+
+    func testNewDeviceCacheInstanceWithNoCachedPurchaserInfoCacheIsStale() {
+        let mockNotificationCenter = MockNotificationCenter()
+        let appUserID = "myUser"
+        self.deviceCache = RCDeviceCache(mockUserDefaults,
+                                         offeringsCachedObject: nil,
+                                         notificationCenter: mockNotificationCenter)
+
+        expect(self.deviceCache.isPurchaserInfoCacheStale(forAppUserID: appUserID,
+                                                          isAppBackgrounded: false)) == true
```
Comment (aboedo) [inline PurchasesTests/Caching/DeviceCacheTests.swift:244]: these would have failed in `develop` since we used to have the `purchaserInfoUpdatedAt` value in memory and not `userDefaults`

File: PurchasesTests/Caching/DeviceCacheTests.swift [hunk 12]
```diff
@@ -194,4 +209,117 @@ class DeviceCacheTests: XCTestCase {
...
+        let appUserID = "myUser"
+        let fourMinutesAgo = Calendar.current.date(byAdding: .minute, value: -4, to: Date())
+        mockUserDefaults.mockValues["com.revenuecat.userdefaults.purchaserInfoLastUpdated.\(appUserID)"] = fourMinutesAgo
+        self.deviceCache = RCDeviceCache(mockUserDefaults,
+                                         offeringsCachedObject: nil,
+                                         notificationCenter: mockNotificationCenter)
+
+        expect(self.deviceCache.isPurchaserInfoCacheStale(forAppUserID: appUserID,
+                                                          isAppBackgrounded: false)) == false
+    }
```
Comment (vegaro) [inline PurchasesTests/Caching/DeviceCacheTests.swift:224]: What's the difference between this test and `testIsPurchaserInfoCacheStaleForForeground`. Also, in this test you're calling `mockUserDefaults.mockValues` but in the other tests you are setting it through `setPurchaserInfoCacheTimestamp`

File: PurchasesTests/Caching/DeviceCacheTests.swift [hunk 13]
```diff
@@ -194,4 +209,117 @@ class DeviceCacheTests: XCTestCase {
...
+        let cacheDateStaleForForeground = Calendar.current.date(byAdding: .hour, value: -23, to: Date())!
+        mockCachedObject.updateCacheTimestamp(with: cacheDateStaleForForeground)
+        expect(self.deviceCache.isOfferingsCacheStale(withIsAppBackgrounded: false)) == true
+        expect(self.deviceCache.isOfferingsCacheStale(withIsAppBackgrounded: true)) == false
+
+        let cacheDateValidForBoth = Calendar.current.date(byAdding: .minute, value: -3, to: Date())!
+        mockCachedObject.updateCacheTimestamp(with: cacheDateValidForBoth)
+        expect(self.deviceCache.isOfferingsCacheStale(withIsAppBackgrounded: false)) == false
+        expect(self.deviceCache.isOfferingsCacheStale(withIsAppBackgrounded: true)) == false
+    }
```
Comment (vegaro) [inline PurchasesTests/Caching/DeviceCacheTests.swift:324]: I feel we need a test for the migration case where the purchaser info timestamp is nil, but the cached purchaser info is not. This can happen for people updating from an older version. I don't think it's covered in any of these tests

File: PurchasesTests/SubscriberAttributes/PurchasesSubscriberAttributesTests.swift [hunk 14]
```diff
@@ -140,15 +140,15 @@ class PurchasesSubscriberAttributesTests: XCTestCase {
 
         setupPurchases()
 
-        expect(self.mockOperationDispatcher.invokedDispatchOnWorkerThreadCount) == 1
         expect(self.mockBackend.invokedGetSubscriberDataCount) == 1
-        expect(self.mockSubscriberAttributesManager.invokedSyncAttributesForAllUsersCount) == 0
+        expect(self.mockDeviceCache.cachePurchaserInfoCount) == 1
```
Comment (vegaro) [inline PurchasesTests/SubscriberAttributes/PurchasesSubscriberAttributesTests.swift:144]: Isn't this line the same as the next one?

File: Purchases/Public/RCPurchases.m [hunk 15]
```diff
@@ -299,7 +299,7 @@ - (instancetype)initWithAppUserID:(nullable NSString *)appUserID
 
         [self.systemInfo isApplicationBackgroundedWithCompletion:^(BOOL isBackgrounded) {
             if (!isBackgrounded) {
-                [self.operationDispatcher dispatchOnWorkerThread:^{
+                [self.operationDispatcher dispatchOnWorkerThreadWithRandomDelay:NO block:^{
```
[the remark this answers is not in the document]
Comment (aboedo) [reply Purchases/Public/RCPurchases.m:302]: yeah, I'll separate into two methods. I'd started by doing that, then I moved to having the randomDelay be optional, then the optional randomDelay looked great in swift but ported very poorly to obj-c, so I killed the optionality. I'll go back to the start.

File: Purchases/Public/RCPurchases.m [hunk 16]
```diff
@@ -863,25 +868,30 @@ - (void)cachePurchaserInfo:(RCPurchaserInfo *)info forAppUserID:(NSString *)appU
...
-                                          } else {
-                                              [self.deviceCache clearPurchaserInfoCacheTimestamp];
-                                          }
-
-                                          CALL_IF_SET_ON_MAIN_THREAD(completion, info, error);
-                                      }];
+    [self.deviceCache setPurchaserInfoCacheTimestampToNowForAppUserID:appUserID];
+    [self.operationDispatcher dispatchOnWorkerThreadWithRandomDelay:isAppBackgrounded block:^{
+        [self.backend getSubscriberDataWithAppUserID:appUserID
+                                          completion:^(RCPurchaserInfo *_Nullable info,
```
[the remark this answers is not in the document]
Comment (aboedo) [reply Purchases/Public/RCPurchases.m:883]: I always forget these. I need an AppCode setting for it before I go nuts

File: Purchases/Public/RCPurchases.m [hunk 17]
```diff
@@ -894,31 +904,34 @@ - (void)performOnEachProductIdentifierInOfferings:(NSDictionary *)offeringsData
...
-        RCDebugLog(@"No cached offerings, fetching");
-        [self updateOfferingsCache:completion];
-    }
+    }];
 }
 
-- (void)updateOfferingsCache:(nullable RCReceiveOfferingsBlock)completion {
+- (void)updateOfferingsCache:(nullable RCReceiveOfferingsBlock)completion isAppBackgrounded:(BOOL)isAppBackgrounded {
     [self.deviceCache setOfferingsCacheTimestampToNow];
-    __weak typeof(self) weakSelf = self;
```
[the remark this answers is not in the document]
Comment (aboedo) [reply Purchases/Public/RCPurchases.m:912]: I don't think we needed it in the first place - I believe I made a mistake in thinking that there was a retention cycle there, but really, the block is used only here and the completion block isn't stored elsewhere, so as soon as the completion is executed, it dies and the block is released, then the instance is ready to be released. 

The case handled here is not trivial, though: 

if you call reset after calling offerings and the callback hasn't returned: 
- if you're doing this weak referencing (current case), the purchases instance is killed, it no longer holds on to the completion block, and completion isn't called. 
- if you're doing this with strong referencing (it's the default, and what we do for almost all other methods), the purchases instance lives until the completion block is called, which might lead to some unexpected behavior if there are any values in memory for the instance that are outdated after the reset. 

I think we should at some point actually audit this behavior and check out what happens after a reset with some of these completion blocks. 

As of right now, I'm not sure completion not returning is better or worse than it returning but potentially messing up if there's anything outdated in memory from the previous purchases instance.

File: PurchasesCoreSwift/Misc/OperationDispatcher.swift [hunk 18]
```diff
@@ -26,8 +27,13 @@ import Foundation
         }
     }
 
-    @objc public func dispatchOnWorkerThread(_ block: @escaping () -> Void) {
-        workerQueue.async { block() }
+    @objc public func dispatchOnWorkerThread(withRandomDelay: Bool = false,
+                                             block: @escaping () -> ()) {
+        if withRandomDelay {
+            let delay = Double.random(in: 0..<maxJitterInSeconds)
```
[the remark this answers is not in the document]
Comment (aboedo) [reply PurchasesCoreSwift/Misc/OperationDispatcher.swift:33]: yep

[the remark this answers is not in the document]
Comment (aboedo) [reply PurchasesCoreSwift/Misc/OperationDispatcher.swift:33]: doubles, so really it'll be like `0.382350321832`. and I'll update the maxJitter to 5 before merging

File: PurchasesTests/Caching/DeviceCacheTests.swift [hunk 19]
```diff
@@ -99,60 +100,74 @@ class DeviceCacheTests: XCTestCase {
...
     }
 
     func testPurchaserInfoIsProperlyCached() {
         let data = Data()
         self.deviceCache.cachePurchaserInfo(data, forAppUserID: "cesar")
+
         expect(self.mockUserDefaults.mockValues["com.revenuecat.userdefaults.purchaserInfo.cesar"] as? Data)
             .to(equal(data))
-        expect(self.deviceCache.cachedPurchaserInfoData(forAppUserID: "cesar")).to(equal(data))
-        expect(self.mockUserDefaults.setObjectForKeyCalledValue)
```
[the remark this answers is not in the document]
Comment (aboedo) [reply PurchasesTests/Caching/DeviceCacheTests.swift:154]: the check on 153 is redundant with the one that's right above it, and the check in 154 actually assumes that the _last_ thing you store is the purchaserInfo, which breaks when you store the purchaserInfoUpdatedDate

File: PurchasesTests/Caching/DeviceCacheTests.swift (hunk 13 above)
Comment (aboedo) [reply PurchasesTests/Caching/DeviceCacheTests.swift:324]: good catch

File: Purchases/Public/RCPurchases.m (hunk 15 above)
[the remark this answers is not in the document]
Comment (aboedo) [reply Purchases/Public/RCPurchases.m:302]: ok, I was trying this and remembered why I ended up going with the param: it's so that callers don't have to choose between methods, since almost every time they pass in the value of `isBackgrounded`. 

So if there are two separate methods, you'd have to do `if isBackgrounded dispatchWithRandomDelay else dispatch`. 
note that we almost every time pass `isBackgrounded`. even in this line, we're passing `NO` because we know that `isBackgrounded` is false, because of the `if` in the line right above. But we could just as easily pass `isBackgrounded` instead.

File: PurchasesTests/SubscriberAttributes/PurchasesSubscriberAttributesTests.swift (hunk 14 above)
Comment (aboedo) [reply PurchasesTests/SubscriberAttributes/PurchasesSubscriberAttributesTests.swift:144]: they're actually different, although the naming could be improved - one is for the `read` method and the other one for the `write`

File: PurchasesTests/Caching/DeviceCacheTests.swift (hunk 12 above)
Comment (aboedo) [reply PurchasesTests/Caching/DeviceCacheTests.swift:224]: I'll take any naming suggestions I can get, but here's what they're supposed to check: 

`testIsPurchaserInfoCacheStaleForForeground`: 
test that the value of isPurchaserInfoCacheStale works correctly when the app is in the foreground. 
- create a cached purchaser info, but stale date. 
- value should be true
- update the purchaserInfo cache date, with a recent value
- value should be false. 


`testNewDeviceCacheInstanceWithExistingValidPurchaserInfoCacheIsntStale` is supposed to ensure that new instances won't return that the value is stale unless it actually is. This is because we previously were storing the lastUpdated value in memory, so this would have had a different result before this PR. 
- store a value for lastUpdated, that's recent
- check that it isn't considered stale

As for using the two mechanisms to mock, they can be unified. I had just forgotten that you could set it directly through the protected method. 
Using both mechanisms might be better coverage? but less consistent. I'm happy to unify.

Comment (vegaro) [reply PurchasesTests/Caching/DeviceCacheTests.swift:224]: it's ok, both mechanisms work. I wasn't sure I was missing somehting. Thanks for the explanation

File: Purchases/Caching/RCDeviceCache.m [hunk 20]
```diff
@@ -128,9 +128,12 @@ - (void)cachePurchaserInfo:(NSData *)data forAppUserID:(NSString *)appUserID {
 
 - (BOOL)isPurchaserInfoCacheStaleForAppUserID:(NSString *)appUserID isAppBackgrounded:(BOOL)isAppBackgrounded {
     NSDate * _Nullable purchaserInfoCachesLastUpdated = [self purchaserInfoCachesLastUpdatedForAppUserID:appUserID];
+    if (!purchaserInfoCachesLastUpdated) {
+        return YES;
+    }
     NSTimeInterval timeSinceLastCheck = -[purchaserInfoCachesLastUpdated timeIntervalSinceNow];
     int cacheDurationInSeconds = [self cacheDurationInSecondsWithIsAppBackgrounded:isAppBackgrounded];
-    return !(purchaserInfoCachesLastUpdated != nil && timeSinceLastCheck < cacheDurationInSeconds);
+    return timeSinceLastCheck >= cacheDurationInSeconds;
```
Comment (vegaro) [inline Purchases/Caching/RCDeviceCache.m:136]: much easier to read

**Review by vegaro — APPROVED** [review]
